### PR TITLE
feat: Parse input to compose file version (RET-394)

### PR DESCRIPTION
**Pull request by DerekMaggio**
# Overview

Add conversion later for converting from input file to docker compose file

# Changelog

- Create ConversionLayer class
- Add mapping from SystemConfigurationModel.compose_file_version to RuntimeComposeFileModel.version
- Add tests for this

# Review requests

Would love some feedback about the structure of my conversion layer.

# Risk assessment

None

### CODE REVIEW

File: emulation_system/emulation_system/compose_file_creator/conversion_layer.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        self._config_model = config_model
+        self.compose_model = RuntimeComposeFileModel()
+
+        self._set_version()
+
+    def _set_version(self) -> None:
+        """Sets version on Compose file."""
+        self.compose_model.version = self._config_model.compose_file_version
+
+    def add_ports(self) -> None:
```
Comment (amitlissack) [inline emulation_system/emulation_system/compose_file_creator/conversion_layer.py:27]: What are the inputs to the `add` methods?

Comment (DerekMaggio) [reply emulation_system/emulation_system/compose_file_creator/conversion_layer.py:27]: These are just placeholders for right now. I am working on them in a separate PR

Comment (DerekMaggio) [reply emulation_system/emulation_system/compose_file_creator/conversion_layer.py:27]: They will actually end up being private methods

File: emulation_system/emulation_system/compose_file_creator/conversion_layer.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        return service
+
+    def _create_services(self) -> None:
+        """Define services for compose file."""
+        self.compose_model.services = {
+            container.id: self._configure_service(container)
+            for container in self._config_model.containers.values()
+        }
+
+    def add_ports(self) -> None:
```
Comment (sfoster1) [inline emulation_system/emulation_system/compose_file_creator/conversion_layer.py:63]: From a style perspective, I tend not to like methods like this in Python. They bring to mind the [builder pattern](https://en.wikipedia.org/wiki/Builder_pattern), a pattern that really took off in java because of the lack of 1) easy to create mapping primitives 2) named arguments - it was basically the only way to have a bag-of-data or configuration-heavy class constructor where you wouldn't have to like have the class def up in another window and be following the positional arguments with your finger.

The downside to the pattern is that it's very hard to programmatically assess when you're done building a class, and what the type of intermediates in the builder chain is. If the way you build this class looks like this:
```
dockerfile = ConversionLayer().add_ports(...).add_env_vars(...).add...
```

What's the type of the intermediates between the chain calls? (and even if you're separating those to different lines, the same thing applies).

Even if those methods are private and called by the initializer, if they're setting private members inside the methods then it's hard for static analysis like mypy to figure out what should be present and when. It also makes it hard to make this bag-of-data class non-mutable.

I'd advocate for instead doing this architectural pattern:

- Have a model for the emulator stack consisting of some set of dataclass types, probably frozen but maybe not. This model is the One True Internal Representation.
- Have adapters or conversion layers that can ingest from / export to docker-compose files, emulator stack config files, etc etc

Basically, put the data and the data model at the center and model the transformers as pure functions - inputs and outputs, no state - that operate on the data model. At that point, things become easier to test, easier to type, and easier to read.

**Review by sfoster1 — CHANGES_REQUESTED** [review]
Uh I put my big comment inline sorry but let's center this code on the data model and have free, pure functions for the translation layer.

File: emulation_system/emulation_system/compose_file_creator/conversion_layer.py (hunk 2 above)
Comment (DerekMaggio) [reply emulation_system/emulation_system/compose_file_creator/conversion_layer.py:63]: > Have adapters or conversion layers that can ingest from / export to docker-compose files, emulator stack config files, etc etc

Does this mean that I should not have intermediary layers?

Comment (DerekMaggio) [reply emulation_system/emulation_system/compose_file_creator/conversion_layer.py]: @sfoster1 check out my latest commit and let me know what you think.

**Review by sfoster1 — APPROVED** [review]
Nice, this looks really solid!
